lib/dialer/retrydialer.go

```diff
@@ -0,0 +1,135 @@
+	// Implementations should honour context deadlines, timeouts, and cancellations (if any).
+	DialUpstream(ctx context.Context, upstream core.Upstream) (forwarder.DuplexConn, error)
+}
+
+type SimpleUpstreamDialer struct{}
+
+func (d SimpleUpstreamDialer) DialUpstream(ctx context.Context, upstream core.Upstream) (forwarder.DuplexConn, error) {
+	dd := net.Dialer{}
+	conn, err := dd.DialContext(ctx, upstream.Network, upstream.Address)
+	if err != nil {
+		return nil, err
+	}
+	switch c := conn.(type) {
+	case *net.TCPConn:
+		return c, nil
+	case *tls.Conn:
+		return c, nil
+	default:
+		_ = conn.Close()
+		return nil, ConnectionTypeUnsupported
+	}
+}
+
+// RetryDialer attempts to dial a candidate Upstream as selected by a
+// configurable DialPolicy. If the dial attempt fails, it informs the policy
+// of the failure and asks the policy for the next candidate upstream.
+// RetryDialer requires a Timeout to be supplied, which is shared across
+// all dial attempts.
+//
+// Multiple goroutines may invoke methods on a RetryDialer simultaneously.
+type RetryDialer struct {
+	Logger      slog.Logger
+	Timeout     time.Duration // Timeout to apply for each DialBestUpstream operation.
+	Policy      DialPolicy
+	InnerDialer UpstreamDialer
+}
+
+func (d *RetryDialer) DialBestUpstream(ctx context.Context, candidates core.UpstreamSet) (core.Upstream, forwarder.DuplexConn, error) {
+	if len(candidates) == 0 {
+		return core.Upstream{}, nil, NoCandidateUpstreams
+	}
+	// TODO use shorter timeout for each of n > 1 dial attempts?
+	dialCtx, cancel := context.WithTimeout(ctx, d.Timeout)
+	defer cancel()
+
+	for {
+		upstream, err := d.Policy.ChooseBestUpstream(candidates)
+		if err != nil {
+			// TODO instead of aborting, we could pause here
+			// (honouring dialCtx timeout) to give policy chance to change its mind.
+			return core.Upstream{}, nil, err
+		}
+		// TODO if policy chose an upstream we tried earlier this call, we could also
+		// pause here (honouring dialCtx timeout) with an exponential backoff.
+
+		conn, err := d.InnerDialer.DialUpstream(dialCtx, upstream)
+		if err != nil {
+			// If we exceeded the dial timeout, then dialCtx.Err() is non-nil
+			if dialCtxErr := dialCtx.Err(); dialCtxErr != nil {
+				d.Logger.Warn(&slog.LogRecord{Msg: "dial timed out", Upstream: &upstream})
+				// We cannot infer much about upstream health in this scenario.
+				// Halt and indicate to caller that we timed out.
+				return core.Upstream{}, nil, dialCtxErr
+			}
+			d.Logger.Warn(&slog.LogRecord{Msg: "dial failed", Upstream: &upstream})
+			d.Policy.DialFailed(upstream, err)
+			continue
+		}
+		d.Logger.Info(&slog.LogRecord{Msg: "dial succeeded", Upstream: &upstream})
+		d.Policy.DialSucceeded(upstream)
+
+		// Wrap & instrument the returned conn to inform the DialPolicy on conn Close.
+		wrappedConn := &CloseNotifyingDuplexConn{
+			DuplexConn: conn,
+			upstream:   upstream,
+			policy:     d.Policy,
+		}
+		return upstream, wrappedConn, nil
```

Comment:
i'm not entirely happy about this, but it does seem to fit the pattern of letting the caller of this function acquire the resource, and then be responsible for releasing the resource

```
conn, err := rd.DialBestUpstream(ctx, upstreams)
if err != nil { return err }
defer conn.Close()
DoSomethingWith(conn)
```

an alternative would be to make this layer, inside `DialBestUpstream`, responsible for closing the conn. we could do that if we just did all the work with the established upstream conn in here as well, e.g. connecting the client conn to this new upstream conn.   an upside of that would be that all these dial policy operations would happen right here in the one function, which would be quite clear.  a downside of that is then this function needs to be concerned with forwarding and perhaps with the client conn - currently it has no idea what the resulting conn is used for.